PCDHA12.py

```diff
@@ -0,0 +1,121 @@
+import os
+from abc import ABC, abstractmethod
+
+import numpy as np
+from Bio import SeqIO
+from Bio.SeqRecord import SeqRecord
+from Bio.SeqUtils import GC
+
+
+# 2.4
+def filter_length(record: SeqRecord, length_bounds: tuple) -> bool:
+    return length_bounds[0] <= len(record.seq) <= length_bounds[1]
+
+
+def filter_gc(record: SeqRecord, gc_bounds: tuple) -> bool:
+    return gc_bounds[0] <= GC(record.seq) <= gc_bounds[1]
+
+
+def filter_quality(record: SeqRecord, quality_threshold: int) -> bool:
+    avg_quality = np.mean(record.letter_annotations["phred_quality"])
+    return avg_quality >= quality_threshold
+
+
+def filter_fastq(input_path: str,
+                 gc_bounds: tuple = (0, 100),
+                 length_bounds: tuple = (0, 2 ** 32),
+                 quality_threshold: int = 0) -> None:
+    """ Reads a FASTQ file, filters sequences based on GC content, sequence
+    length and quality threshold, and writes the filtered sequences to
+    a new file """
+    path, filename = os.path.split(input_path)
+    name, ext = os.path.splitext(filename)
+    output_path = path + "/" + f"{name}_filtered{ext}"
+    input_seq_iterator = SeqIO.parse(input_path, 'fastq')
+    filtered_seq_iterator = (record for record in input_seq_iterator
+                             if filter_length(record, length_bounds)
+                             and filter_gc(record, gc_bounds)
+                             and filter_quality(record, quality_threshold))
+    SeqIO.write(filtered_seq_iterator, output_path, "fastq")
+
+
+# 2.5
+class BiologicalSequence(ABC, str):
+    @abstractmethod
+    def check_alphabet(self) -> bool:
+        pass
+
+
+class NucleicAcidSequence(BiologicalSequence):
+    def __init__(self, sequence):
+        raise NotImplementedError('An instance of this class cannot be created')
```

Comment:
Здесь можно было бы при ините делать последовательность сразу большими буквами, чтобы потом в словаре не хранить и маленькие, и большие буквы. И в целом инит вынести в этот класс, и убрать его в дочерних, с точки зрения информации подаваемой, что АА, что РНК, что ДНК это все строчки с буковками. Так что я бы вместо ошибки добавила вот это, а потом от этого дальше наследовалась, чтобы не дублироваться.
```suggestion
    def __init__(self, sequence):
        self.seq = seq.upper()
```